radix-engine/src/engine/modules/execution_trace.rs

```diff
@@ -25,30 +27,225 @@ pub enum VaultOp {
+}
+
+#[derive(Clone, Debug, TypeId, Encode, Decode, PartialEq, Eq)]
+pub struct ProofSnapshot {
+    pub resource_address: ResourceAddress,
+    pub resource_type: ResourceType,
+    pub restricted: bool,
+    pub total_locked: LockedAmountOrIds,
+}
+
+impl From<&ProofSubstate> for ProofSnapshot {
+    fn from(proof: &ProofSubstate) -> ProofSnapshot {
+        ProofSnapshot {
+            resource_address: proof.resource_address,
+            resource_type: proof.resource_type,
+            restricted: proof.restricted,
+            total_locked: proof.total_locked.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub struct TracedSysCallData {
+    pub buckets: HashMap<BucketId, Resource>,
+    pub proofs: HashMap<ProofId, ProofSnapshot>,
+}
+
+impl TracedSysCallData {
+    pub fn new_empty() -> Self {
+        Self {
+            buckets: HashMap::new(),
+            proofs: HashMap::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.buckets.is_empty() && self.proofs.is_empty()
+    }
+}
+
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub struct SysCallTrace {
+    pub origin: SysCallTraceOrigin,
+    pub sys_call_depth: usize,
+    pub call_frame_actor: REActor,
+    pub call_frame_depth: usize,
+    pub instruction_index: Option<usize>,
+    pub input: TracedSysCallData,
+    pub output: TracedSysCallData,
+    pub children: Vec<SysCallTrace>,
+}
+
+#[derive(Debug, Clone, TypeId, Encode, Decode, Eq, PartialEq)]
+pub enum SysCallTraceOrigin {
+    ScryptoFunction(ScryptoFunctionIdent),
+    ScryptoMethod(ScryptoMethodIdent),
+    NativeFunction(NativeFunction),
+    NativeMethod(NativeMethod),
+    CreateNode,
+    DropNode,
+    /// Anything else that isn't traced on its own, but the trace exists for its children
+    Opaque,
+}
+
+pub trait Traceable: Debug {
+    fn buckets(&self) -> Vec<BucketId>;
+    fn proofs(&self) -> Vec<ProofId>;
+}
+
+impl Traceable for ScryptoValue {
+    fn buckets(&self) -> Vec<BucketId> {
+        self.bucket_ids.keys().map(|k| k.clone()).collect()
+    }
+
+    fn proofs(&self) -> Vec<ProofId> {
+        self.proof_ids.keys().map(|k| k.clone()).collect()
+    }
+}
+
+impl Traceable for Bucket {
+    fn buckets(&self) -> Vec<BucketId> {
+        vec![self.0]
+    }
+
+    fn proofs(&self) -> Vec<ProofId> {
+        vec![]
+    }
+}
+
+impl Traceable for Proof {
+    fn buckets(&self) -> Vec<BucketId> {
+        vec![]
+    }
+
+    fn proofs(&self) -> Vec<ProofId> {
+        vec![self.0]
+    }
+}
+
+impl Traceable for (ResourceAddress, Option<Bucket>) {
+    fn buckets(&self) -> Vec<BucketId> {
+        match &self.1 {
+            Some(bucket) => vec![bucket.0.clone()],
+            None => vec![],
+        }
+    }
+
+    fn proofs(&self) -> Vec<ProofId> {
+        vec![]
+    }
+}
+
+impl Traceable for CallFrameUpdate {
+    fn buckets(&self) -> Vec<BucketId> {
+        self.nodes_to_move
```

Comment:
not sure if we also want to trace the refs (`node_refs_to_copy`) here? :thinking: